### PR TITLE
add support omapi options for isc-dhcp

### PR DESCRIPTION
**Pull request by pronix**
should be fix this ticket
http://projects.theforeman.org/issues/3067

### CODE REVIEW

**Comment by aleksf** [comment]
Hi Dmitry,

Thanks for working on this. There's a number of issues with this patch that I picked up during test: 
1. SPARC build test case fails with the error in smart-proxy logfile
   
   E, [2013-12-04T11:56:38.732935 #3807] ERROR -- : DHCP implementation does not support Vendor Options
   
   This can be fixed by working out the conditional in lib/proxy/dhcp/server.rb that makes assumption of unsupported vendor options in ISC DHCP
2. "if options[:vendor] =~ /sun|solar/i" conditional in the patch never gets validated
   
   I think conditional should be based on the architecture of the Host that we are provisioning
3. Working around issues 2 and 3 above, the code populates following lease object, which is missing all SUNW.\* options as in described issue http://projects.theforeman.org/issues/3067
   
   host sparc-host.fqdn {
   dynamic;
   hardware ethernet 00:21:28:6d:62:e8;
   fixed-address 10.229.11.38;
       supersede server.filename = "Solaris-5.10-hw0811-sun4v-inetboot";
       supersede server.next-server = 0a:e5:0b:18;
       supersede host-name = "sparc-host.fqdn";
       supersede server.vendor-option-space = "SUNW";
   }
   
   Options, specified in patch, are passed to OMAPI, however they contain empty values, therefore they are ignored by OMAPI and are not added to the lease object. 
   All SUNW.\* option values must be derived from the Medium (Mirror) object that we select for the SPARC Host that we are provisioning. 
   Please refer to the following OMAPI options below where examples are given based on configured SPARC Medium (Mirror)
   
   Name: Solaris Media
   Path: http://js-server/mrepo/Solaris/install/OS/Solaris_$major.$minor_$arch_$release
   NFS media path: js-server:/Solaris/install/Solaris_$major.$minor_$arch_$release
   NFS configuration path: js-server:/Solaris/jumpstart
   NFS image path: js-server:/Solaris/images/ (this is not used in my testcase, as I don't use Flash Archives)
   Operating System Family: Solaris
   
   option server.vendor-option-space = "SUNW";
   This is working already as it is hardcoded value, however please refer to point 2 above in regards the conditional block.
   
   option server.next-server = 0a:e5:0b:18; 
   This is working as this is the part of standard PXE. It points to Smart-Proxy IP in hex (tftp)
   
   option server.filename = "Solaris-5.10-hw0811-sun4v-inetboot"; 
   This is working as this is the part of standard PXE and specifies network loader image name
   
   option SUNW.JumpStart-server = " ";
   This is not working. Value must be derived from "NFS configuration path" in defined SPARC Medium (Mirror). 
   Value example: "js-server:/Solaris/jumpstart"
   
   option SUNW.sysid-config-file-server = " "; 
   This is not working. Value must be the combination of SUNW.JumpStart-server value and hardcoded path to sysidcfg.
   Value example: "js-server:/Solaris/jumpstart/sysidcfg/sysidcfg_primary"
   
   option SUNW.install-server-hostname = ""; 
   This is not working. Value must be derived from SPARC Medium (Mirror) Media Path.
   Value example: "js-server"
   
   option SUNW.install-server-ip-address = ;  
   This is not working. Value must be derived from SPARC Medium (Mirror) Media Path hostname part.
   Value example: 10.229.11.24 (IP of js-server). It is then decoded to hex by OMAPI.
   
   option SUNW.install-path =""; 
   This is not working. Value must be derived from NFS Media Path based on chosen minor/major release numbers for the host that we are provisioning. 
   Value example: "/Solaris/install/Solaris_5.10_sparc_hw0811"
   
   option SUNW.root-server-hostname = ""; 
   This is not working. Value must be the same as the value of SUNW.install-server-hostname.
   
   option SUNW.root-server-ip-address = ; 
   This is not working. Value must be the same as the value of SUNW.install-server-ip-address
   
   option SUNW.root-path-name = ""; 
   This is nor working. Value must be the same as SUNW.install-path + "/Solaris_10/Tools/Boot" directory
   Value example: "/Solaris/install/Solaris_5.10_sparc_hw0811/Solaris_10/Tools/Boot"

So for every SPARC host, that we are defining. OMAPI option values must be populated from the selected SPARC Mirror. 

Regards,
Alex

**Comment by pronix** [comment]
1. fixed
2. foreman will sent 'vendor' options only for solaris os defined here foreman/app/models/operatingsystems/solaris.rb

**Comment by aleksf** [comment]
Hi Dmitri,

I tested this patch against Foreman 1.3.0-1 base it and it works properly. The only gotcha is, if FQDN is configured in NFS medium for Solaris, it breaks JumpStart option SUNW.JumpStart-server, as it expects either IP or shortname. This can be worked around by defining all hostnames in NFS medium to be shortnames. 

Regards,
Alex

**Comment by domcleal** [comment]
[test]

@aleksf do you think the patch should just strip off the domain from the medium hostname?

**Comment by aleksf** [comment]
@domcleal sorry, option that causes problems is SUNW.install-server-hostname that takes its value from app/models/operatingsystems/solaris.rb from server_name in jumpstart_params function in Foreman

Will create new pull request to strip the domain off from host.medium.media_host.

I've tested in my environment and it works.

**Comment by aleksf** [comment]
@domcleal foreman pull request to split domain from medium hostname is https://github.com/theforeman/foreman/pull/1084

**Comment by pronix** [comment]
hello @domcleal 
can you this pull request merge or require something change ?

**Comment by ohadlevy** [comment]
[test]

File: lib/proxy/dhcp/server/isc.rb [hunk 1]
```diff
@@ -38,6 +38,40 @@ def addRecord options = {}
       statements << bootServer(options[:nextServer])             if options[:nextServer]
       statements << "option host-name = \\\"#{record.name}\\\";" if record.name
 
+      # omapi
```
Comment (ohadlevy) [inline lib/proxy/dhcp/server/isc.rb]: please extract this all to a seperate method, something like

``` ruby
  statements += solaris_options_statements(options)
```

and then define another method, something like

``` ruby
    def solaris_options_statements(options)
      # Solaris options defined in Foreman app/models/operatingsystems/solaris.rb method jumpstart_params
      # options example
      # {"hostname"                                      => ["itgsyddev910.macbank"],
      #  "mac"                                           => ["00:21:28:6d:62:e8"],
      #  "ip"                                            => ["10.229.11.38"],
      #  "network"                                       => ["10.229.11.0"],
      #  "nextServer"                                    => ["10.229.11.24"], "filename" => ["Solaris-5.10-hw0811-sun4v-inetboot"],
      #  "<SPARC-Enterprise-T5120>root_path_name"        => ["/Solaris/install/Solaris_5.10_sparc_hw0811/Solaris_10/Tools/Boot"],
      #  "<SPARC-Enterprise-T5120>sysid_server_path"     => ["10.229.11.24:/Solaris/jumpstart/sysidcfg/sysidcfg_primary"],
      #  "<SPARC-Enterprise-T5120>install_server_ip"     => ["10.229.11.24"],
      #  "<SPARC-Enterprise-T5120>jumpstart_server_path" => ["10.229.11.24:/Solaris/jumpstart"],
      #  "<SPARC-Enterprise-T5120>install_server_name"   => ["itgsyddev807.macbank"],
      #  "<SPARC-Enterprise-T5120>root_server_hostname"  => ["itgsyddev807.macbank"],
      #  "<SPARC-Enterprise-T5120>root_server_ip"        => ["10.229.11.24"],
      #  "<SPARC-Enterprise-T5120>install_path"          => ["/Solaris/install/Solaris_5.10_sparc_hw0811"] }
      #
      statements = []
      options.each do |key, value|
```

File: lib/proxy/dhcp/server/isc.rb [hunk 2]
```diff
@@ -125,6 +159,24 @@ def parse_record_options text
         # OMAPI settings
       when /^deleted/
         options[:deleted] = true
+      when 'vendor-option-space SUNW'
```
Comment (ohadlevy) [inline lib/proxy/dhcp/server/isc.rb:167]: ideally do the same here?

File: test/server_isc_test.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+ENV['RACK_ENV'] = 'test'
+
+class ServerIscTest < Test::Unit::TestCase
+  include Rack::Test::Methods
+  def setup
+    user     ||= "user"
+    pass     ||= "pass"
+    @host    ||= "host"
+    provider ||= "ipmitool"
```
Comment (ohadlevy) [inline test/server_isc_test.rb]: copy paste?

File: test/server_isc_test.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      "<SPARC-Enterprise-T5120>root_server_hostname"=>"itgsyddev807.macbank",
+      "<SPARC-Enterprise-T5120>root_server_ip"=>"192.168.122.24",
+      "<SPARC-Enterprise-T5120>install_path"=>"/Solaris/install/Solaris_5.10_sparc_hw0811"
+    }
+    #Proxy::DHCP::Server::ISC.any_instance.stubs(:read_config).returns({})
+    s=Proxy::DHCP::Server.new('192.168.122.1')
+    sub=Proxy::DHCP::Subnet.new(s,'192.168.122.0','255.255.255.0')
+    Proxy::DHCP::Server::ISC.any_instance.stubs(:find_subnet).returns(sub)
+    post '/dhcp/192.168.122.10', data
+    assert last_response.ok?, 'Last response was not ok'
```
Comment (ohadlevy) [inline test/server_isc_test.rb]: not ok?

**Comment by ohadlevy** [comment]
also, do we have a redmine ticket for this? 
thanks!

File: test/server_isc_test.rb (hunk 3 above)
Comment (pronix) [inline test/server_isc_test.rb]: it is copypasted from other tests in foreman proxy, but customized for checking omapi options

**Comment by pronix** [comment]
ticket is http://projects.theforeman.org/issues/3067

File: lib/proxy/dhcp/server/isc.rb (hunk 2 above)
Comment (pronix) [inline lib/proxy/dhcp/server/isc.rb:167]: sorry, i not understand what you mean.
thanks

Comment (ohadlevy) [inline lib/proxy/dhcp/server/isc.rb:167]: It make sense to split into a separate method, so each vendor option gets its own method for parsing the arguments

File: test/server_isc_test.rb (hunk 3 above)
Comment (ohadlevy) [inline test/server_isc_test.rb]: i mean the provider line, probably not required.

File: lib/proxy/dhcp/server/isc.rb (hunk 2 above)
Comment (pronix) [inline lib/proxy/dhcp/server/isc.rb:167]: do you think it will more easy for read ?
current case..when for me most clear.

Comment (ohadlevy) [inline lib/proxy/dhcp/server/isc.rb:167]: I'll send you an example

Comment (pronix) [inline lib/proxy/dhcp/server/isc.rb:167]: thanks.

Comment (ohadlevy) [inline lib/proxy/dhcp/server/isc.rb:167]: see https://github.com/ohadlevy/smart-proxy/commit/7980009294358567533b4183014dd71790bcb74d

**Comment by domcleal** [comment]
Thanks for the contribution @pronix, I've merged it in as 0092c5d with @ohadlevy's method refactor.  This will be released in Foreman 1.4.

**Comment by pronix** [comment]
thanks.

**Pull request closed without merging**